File: app/src/lib/erinnerungen.ts
/**
 * Erinnerungen: zu einer Zeit gross auf dem Schirm, bis jemand bestätigt.
 *
 * Der Hub ist nur die geteilte Ablage (Familien-Sammlung «reminders») -
 * ob eine Erinnerung fällig ist, rechnet jedes Gerät selbst aus seiner
 * Uhr. So braucht es keinen Wecker im Hub, und das Wandpanel zeigt das
 * Vollbild auch dann pünktlich, wenn die Verbindung gerade stockte.
 * Bestätigen schreibt `done` in die Ablage; der Hub meldet die Änderung
 * allen offenen Apps, und das Vollbild verschwindet überall.
 */

export interface Erinnerung {
  id: string;
  text?: unknown;
  /** Zielzeitpunkt in Millisekunden seit 1970. */
  at?: unknown;
  /** Bestätigt - taucht nirgends mehr auf. */
  done?: unknown;
  /** Gross auf den Bildschirmen zeigen. Fehlt das Feld, gilt ja -
   *  Einträge von vor den Schaltern kannten nur diesen Weg. */
  anzeigen?: unknown;
  /** Zur Zeit eine Push-Nachricht schicken - verschickt der Hub. */
  push?: unknown;
  /** An wen der Push geht; leer heisst alle im Haushalt. */
  push_an?: unknown;
  /** Wer die Erinnerung nur für sich weggedrückt hat («Erledigt» statt
   *  «Für alle erledigt»). Bei diesen Benutzern bleibt das Vollbild weg,
   *  bei allen anderen steht es weiter - erst `done` räumt überall ab. */
  quittiert?: unknown;
  /** Wiederholung («daily», «weekly», «monthly», «yearly») - dieselben
   *  Schlüssel wie bei den Aufgaben. Fehlt das Feld: einmalig. Beim
   *  Bestätigen wird nicht erledigt, sondern auf den nächsten Termin
   *  weitergestellt. */
  repeat?: unknown;
  /** Vom Hub gesetzt: Der Push zu diesem Termin ist verschickt. Beim
   *  Weiterstellen wird es zurückgenommen, damit der nächste rausgeht. */
  pushed?: unknown;
}

/** Die Wiederholungen, die eine Erinnerung kennt - Schlüssel wie bei
 *  den Aufgaben (REPEAT_OPTIONS), plus jährlich: den Heizungs-Service
 *  gibt es, den täglichen Jahresputz nicht. */
export const WIEDERHOLUNGEN = [
  { key: 'none', label: 'einmalig' },
  { key: 'daily', label: 'täglich' },
  { key: 'weekly', label: 'wöchentlich' },
  { key: 'monthly', label: 'monatlich' },
  { key: 'yearly', label: 'jährlich' },
] as const;

export type Wiederholung = 'daily' | 'weekly' | 'monthly' | 'yearly';

/** Die Wiederholung eines Eintrags - oder null für einmalig (rein, testbar). */
export function wiederholungVon(eintrag: Erinnerung): Wiederholung | null {
  const roh = String(eintrag.repeat ?? '');
  return roh === 'daily' || roh === 'weekly' || roh === 'monthly' || roh === 'yearly'
    ? roh
    : null;
}

/** Das deutsche Wort zur Wiederholung - für Listen und Badges. */
export function wiederholungsLabel(wiederholung: Wiederholung): string {
  return (
    WIEDERHOLUNGEN.find((option) => option.key === wiederholung)?.label ?? wiederholung
  );
}

/** Der nächste Termin nach «jetzt» (rein, testbar).
 *
 *  Gerechnet wird im Kalender, nicht in Millisekunden: 7 Uhr bleibt
 *  7 Uhr, auch über die Zeitumstellung hinweg. Monat und Jahr zählen
 *  vom **ursprünglichen** Termin aus - eine Erinnerung vom 31. rutscht
 *  im kurzen Monat auf dessen letzten Tag, kehrt danach aber auf den
 *  31. zurück, statt für immer beim 28. zu bleiben. `null`, wenn kein
 *  Termin gefunden wird (kaputter Zeitpunkt).
 */
export function naechsteFaelligkeit(
  atMs: number,
  wiederholung: Wiederholung,
  jetztMs: number
): number | null {
  if (!Number.isFinite(atMs) || atMs <= 0) return null;
  const start = new Date(atMs);
  for (let schritt = 1; schritt <= 5000; schritt++) {
    const wann = new Date(start);
    if (wiederholung === 'daily') {
      wann.setDate(start.getDate() + schritt);
    } else if (wiederholung === 'weekly') {
      wann.setDate(start.getDate() + 7 * schritt);
    } else {
      // Erst auf den Monatsersten, dann den Monat stellen, dann den
      // Tag begrenzen - sonst schöbe der 31. den Monat still weiter.
      wann.setDate(1);
      wann.setMonth(start.getMonth() + (wiederholung === 'monthly' ? schritt : 0));
      if (wiederholung === 'yearly') wann.setFullYear(start.getFullYear() + schritt);
      const tage = new Date(wann.getFullYear(), wann.getMonth() + 1, 0).getDate();
      wann.setDate(Math.min(start.getDate(), tage));
    }
    if (wann.getTime() > jetztMs) return wann.getTime();
  }
  return null;
}

/** Wer diese Erinnerung für sich weggedrückt hat (rein, testbar). */
export function quittiertVon(eintrag: Erinnerung): string[] {
  const namen = eintrag.quittiert;
  if (!Array.isArray(namen)) return [];
  return namen.map((name) => String(name)).filter((name) => name.trim() !== '');
}

/** Gehört dieser Eintrag auf die Bildschirme? (rein, testbar) */
export function zeigtAn(eintrag: Erinnerung): boolean {
  return eintrag.anzeigen !== false;
}

/** Was «Für alle erledigt» in die Ablage schreibt (rein, testbar).
 *
 *  Einmalige Erinnerungen werden erledigt. Wiederkehrende werden auf
 *  den nächsten Termin weitergestellt - und dabei frisch: Niemand hat
 *  die neue schon weggedrückt, kein Push ist für sie verschickt.
 *  Findet sich kein nächster Termin (kaputter Zeitpunkt), wird auch
 *  eine wiederkehrende erledigt - für immer offen stehen darf nichts. */
export function bestaetigung(
  eintrag: Erinnerung,
  jetztMs: number
): Record<string, unknown> {
  const wiederholung = wiederholungVon(eintrag);
  if (wiederholung) {
    const next = naechsteFaelligkeit(Number(eintrag.at), wiederholung, jetztMs);
    if (next !== null) return { at: next, quittiert: [], pushed: false };
  }
  return { done: true };
}

/** «TT.MM.JJJJ» + «HH:MM» → Zeitpunkt in ms (rein, testbar).
 *
 *  `null`, wenn eine der Angaben keine ist - auch für den 31.02.: Ein
 *  Datum, das der Kalender nicht kennt, würde JavaScript stumm in den
 *  März verschieben, und die Erinnerung käme einen Monat zu früh oder
 *  gar nicht. */
export function zeitpunkt(datum: string, zeit: string): number | null {
  const d = /^(\d{1,2})\.(\d{1,2})\.(\d{4})$/.exec(String(datum ?? '').trim());
  const z = /^(\d{1,2})[:.](\d{2})$/.exec(String(zeit ?? '').trim());
  if (!d || !z) return null;
  const tag = Number(d[1]);
  const monat = Number(d[2]);
  const jahr = Number(d[3]);
  const stunde = Number(z[1]);
  const minute = Number(z[2]);
  if (stunde > 23 || minute > 59) return null;
  const wann = new Date(jahr, monat - 1, tag, stunde, minute);
  if (
    wann.getFullYear() !== jahr ||
    wann.getMonth() !== monat - 1 ||
    wann.getDate() !== tag
  ) {
    return null;
  }
  return wann.getTime();
}

/** Der Zeitpunkt eines Eintrags als Zahl - oder null für Unlesbares. */
function wann(eintrag: Erinnerung): number | null {
  const at = Number(eintrag.at);
  return Number.isFinite(at) && at > 0 ? at : null;
}

/** Alle offenen, nach Zeit sortiert (rein, testbar). */
export function offene(liste: Erinnerung[] | undefined): Erinnerung[] {
  return (liste ?? [])
    .filter((eintrag) => !eintrag.done && wann(eintrag) !== null)
    .sort((a, b) => (wann(a) ?? 0) - (wann(b) ?? 0));
}

/** Was jetzt gross auf den Schirm gehört (rein, testbar).
 *
 *  Fällig heisst: Zeitpunkt erreicht und noch nicht bestätigt. Es gibt
 *  keine Verfallszeit - wer das Display erst abends sieht, soll die
 *  Erinnerung von mittags noch vorfinden. Genau dafür ist das
 *  Bestätigen da. */
export function faellige(
  liste: Erinnerung[] | undefined,
  jetztMs: number
): Erinnerung[] {
  return offene(liste).filter((eintrag) => (wann(eintrag) ?? Infinity) <= jetztMs);
}

/** Fällig UND fürs Vollbild bestimmt (rein, testbar).
 *
 *  Ein Eintrag, der nur pusht, gehört nicht auf den Schirm - den
 *  erledigt der Hub nach dem Versand von selbst. Und wer die Erinnerung
 *  schon für sich weggedrückt hat (`benutzer` steht in `quittiert`),
 *  bekommt sie nicht noch einmal - die anderen sehr wohl. */
export function anzuzeigende(
  liste: Erinnerung[] | undefined,
  jetztMs: number,
  benutzer?: string | null
): Erinnerung[] {
  return faellige(liste, jetztMs)
    .filter(zeigtAn)
    .filter(
      (eintrag) => !benutzer || !quittiertVon(eintrag).includes(benutzer)
    );
}

/** Wann das nächste Vollbild ansteht - für den Prüf-Takt (rein, testbar).
 *
 *  `null` heisst: nichts offen, kein Takt nötig. Der Takt selbst bleibt
 *  grob (die App prüft ohnehin regelmässig); diese Zahl sagt nur, ob
 *  sich das Ticken überhaupt lohnt. */
export function naechsteAt(liste: Erinnerung[] | undefined): number | null {
  const erste = offene(liste)[0];
  return erste ? wann(erste) : null;
}

/** Die Wochen eines Monats als Raster (rein, testbar).
 *
 *  Für den Datums-Wähler: Jede Woche eine Zeile Montag-Sonntag, leere
 *  Plätze vor dem Ersten und nach dem Letzten als null. Montag zuerst -
 *  so hängen die Kalender in diesem Haushalt.
 */
export function monatsraster(jahr: number, monat: number): (number | null)[][] {
  const erster = new Date(jahr, monat - 1, 1);
  const tage = new Date(jahr, monat, 0).getDate();
  // getDay(): 0 = Sonntag. Auf Montag = 0 gedreht.
  const versatz = (erster.getDay() + 6) % 7;
  const zellen: (number | null)[] = [
    ...Array.from({ length: versatz }, () => null),
    ...Array.from({ length: tage }, (_, i) => i + 1),
  ];
  while (zellen.length % 7 !== 0) zellen.push(null);
  const wochen: (number | null)[][] = [];
  for (let i = 0; i < zellen.length; i += 7) wochen.push(zellen.slice(i, i + 7));
  return wochen;
}

/** Monat blättern: 13 wird Januar des Folgejahres (rein, testbar). */
export function monatsSprung(
  jahr: number,
  monat: number,
  schritt: number
): { jahr: number; monat: number } {
  const roh = new Date(jahr, monat - 1 + schritt, 1);
  return { jahr: roh.getFullYear(), monat: roh.getMonth() + 1 };
}

File: hub/homepilot/core/erinnerungen.py
"""Erinnerungen: der Hub schickt den Push, die Bildschirme zeigen selbst.

Die Erinnerungen liegen als Familien-Sammlung in ``family_reminders``
(siehe api/routes/family.py); ob eine **am Bildschirm** fällig ist,
rechnet jedes Gerät aus seiner eigenen Uhr. Für den Push geht das
nicht: Er muss auch dann kommen, wenn gerade keine App offen ist. Also
führt der Hub genau dafür einen kleinen Takt - keine Wecker-Verwaltung,
nur ein Blick auf die Liste alle paar Sekunden.

Ein Eintrag trägt zwei Schalter aus der App: ``anzeigen`` (gross auf
den Schirmen, bis die erste Person bestätigt) und ``push`` samt
``push_an`` (an welche Haushaltsmitglieder). ``pushed`` setzt der Hub
nach dem Versand - in die Sammlung selbst, damit ein Neustart nicht
noch einmal schickt.

Dazu kann ein Eintrag ``repeat`` tragen (daily/weekly/monthly/yearly):
Dann wird er nie erledigt, sondern weitergestellt - vom Bestätigen in
der App, oder bei reinen Push-Erinnerungen vom Hub nach dem Versand.
"""

from __future__ import annotations

import asyncio
import calendar
import logging
import time
from datetime import datetime, timedelta
from typing import Any

log = logging.getLogger(__name__)

#: Der Blick auf die Liste. Grob mit Absicht: Eine Erinnerung ist auf
#: die Minute gestellt, nicht auf die Sekunde.
TAKT_SEKUNDEN = 15.0


def zu_pushen(rows: Any, jetzt_ms: float) -> list[dict[str, Any]]:
    """Welche Einträge jetzt einen Push brauchen (rein, testbar).

    ``at`` steht in Millisekunden - die App rechnet in ``Date.now()``,
    und der Hub übernimmt die Einheit, statt an einer Umrechnung zu
    scheitern, die niemand sieht.
    """
    faellig = []
    for row in rows or []:
        if not isinstance(row, dict) or not row.get("id"):
            continue
        if not row.get("push") or row.get("done") or row.get("pushed"):
            continue
        try:
            at = float(row.get("at") or 0)
        except (TypeError, ValueError):
            continue
        if 0 < at <= jetzt_ms:
            faellig.append(row)
    return faellig


def empfaenger(row: dict[str, Any]) -> list[str]:
    """Die gewählten Namen - leer heisst «alle» (rein, testbar)."""
    namen = row.get("push_an")
    if not isinstance(namen, list):
        return []
    return [str(name).strip() for name in namen if str(name).strip()]


#: Wiederholungen, die eine Erinnerung kennt - dieselben Schlüssel wie
#: in der App (lib/erinnerungen.ts) und bei den Familien-Aufgaben.
WIEDERHOLUNGEN = ("daily", "weekly", "monthly", "yearly")


def naechste_faelligkeit(
    at_ms: float, wiederholung: str, jetzt_ms: float
) -> float | None:
    """Der nächste Termin nach «jetzt» (rein, testbar).

    Das Spiegelbild von ``naechsteFaelligkeit`` in lib/erinnerungen.ts -
    beide Seiten müssen denselben Termin ausrechnen, sonst stellt der
    Hub eine Erinnerung anders weiter als die App. Gerechnet wird im
    Kalender, nicht in Millisekunden: 7 Uhr bleibt 7 Uhr, auch über die
    Zeitumstellung. Monat und Jahr zählen vom **ursprünglichen** Termin
    aus - der 31. rutscht im kurzen Monat auf dessen letzten Tag und
    kehrt danach zurück. Verpasste Termine werden übersprungen, nicht
    nachgeholt.
    """
    if wiederholung not in WIEDERHOLUNGEN:
        return None
    try:
        start = datetime.fromtimestamp(float(at_ms) / 1000)
    except (TypeError, ValueError, OSError, OverflowError):
        return None
    for schritt in range(1, 5001):
        if wiederholung == "daily":
            wann = start + timedelta(days=schritt)
        elif wiederholung == "weekly":
            wann = start + timedelta(days=7 * schritt)
        else:
            jahr = start.year + (schritt if wiederholung == "yearly" else 0)
            monat = start.month + (schritt if wiederholung == "monthly" else 0)
            jahr += (monat - 1) // 12
            monat = (monat - 1) % 12 + 1
            tag = min(start.day, calendar.monthrange(jahr, monat)[1])
            wann = start.replace(year=jahr, month=monat, day=tag)
        ergebnis = wann.timestamp() * 1000
        if ergebnis > jetzt_ms:
            return ergebnis
    return None


def nach_versand(
    rows: Any, ids: set[str], jetzt_ms: float | None = None
) -> list[dict[str, Any]]:
    """Die Liste nach dem Versand (rein, testbar).

    ``pushed`` hält den Versand fest. Ein Eintrag, der **nur** pusht
    (``anzeigen`` aus), ist damit erledigt - es gibt keinen Bildschirm,
    auf dem ihn jemand bestätigen könnte, und offen stehen bliebe er
    sonst für immer. Ausser er wiederholt sich: Dann wird er auf den
    nächsten Termin weitergestellt, frisch (kein ``pushed``, kein
    ``quittiert``), damit der nächste Push wieder rausgeht. Bei
    Einträgen **mit** Bildschirm stellt erst das Bestätigen weiter -
    die App rechnet dort denselben Termin aus (bestaetigung()).
    """
    neue = []
    for row in rows or []:
        if isinstance(row, dict) and str(row.get("id")) in ids:
            wiederholung = str(row.get("repeat") or "")
            if not row.get("anzeigen", True):
                naechste = naechste_faelligkeit(
                    row.get("at") or 0,
                    wiederholung,
                    jetzt_ms if jetzt_ms is not None else time.time() * 1000,
                )
                if naechste is not None:
                    row = {**row, "at": naechste, "pushed": False, "quittiert": []}
                else:
                    row = {**row, "pushed": True, "done": True}
            else:
                row = {**row, "pushed": True}
        neue.append(row)
    return neue


def benutzer_umbenennen(rows: Any, alt: str, neu: str) -> list[Any]:
    """Einen Benutzernamen in offenen Erinnerungen nachziehen (rein, testbar).

    In ``push_an`` (wer den Push bekommt) und ``quittiert`` (wer schon
    für sich bestätigt hat) stehen Namen. Bleibt dort der alte, geht
    der Push an niemanden, und die Erinnerung erscheint wieder, obwohl
    man sie weggedrückt hat.
    """
    neue = []
    for row in rows or []:
        if isinstance(row, dict):
            geaendert = dict(row)
            for feld in ("push_an", "quittiert"):
                namen = geaendert.get(feld)
                if isinstance(namen, list) and alt in namen:
                    geaendert[feld] = [neu if name == alt else name for name in namen]
            row = geaendert
        neue.append(row)
    return neue


async def push_loop(hub: Any) -> None:
    """Alle TAKT_SEKUNDEN: fällige Pushes verschicken und festhalten."""
    while True:
        await asyncio.sleep(TAKT_SEKUNDEN)
        try:
            await _runde(hub)
        except asyncio.CancelledError:
            raise
        except Exception:
            # Ein Fehler im Takt darf den Takt nicht beenden - sonst
            # bleibt jede weitere Erinnerung stumm, und niemand merkt es.
            log.exception("Erinnerungs-Push fehlgeschlagen")


async def _runde(hub: Any) -> None:
    rows = hub.data.get("family_reminders")
    faellig = zu_pushen(rows, time.time() * 1000)
    if not faellig:
        return
    for row in faellig:
        namen = empfaenger(row)
        tokens: list[str] = []
        # Je Person einzeln aufgelöst: recipients() kennt «alle», Rollen
        # und Namen - hier sind es Namen, und die Vereinigung vermeidet
        # doppelte Geräte, wenn jemand zweimal gewählt wurde.
        for ziel in namen or ["all"]:
            for token in hub.push.recipients(hub.users.users, to=ziel):
                if token not in tokens:
                    tokens.append(token)
        text = str(row.get("text") or "Erinnerung")
        result = await hub.push.send(tokens, "⏰ Erinnerung", text)
        log.info(
            "Erinnerung «%s» als Push an %s (%d Geräte angenommen)",
            text,
            ", ".join(namen) or "alle",
            getattr(result, "accepted", 0) or 0,
        )
    ids = {str(row.get("id")) for row in faellig}
    hub.data.set("family_reminders", nach_versand(hub.data.get("family_reminders"), ids))
    # Der Fingerzeig an die offenen Apps - ein nur-Push-Eintrag ist
    # jetzt erledigt und soll aus den Listen verschwinden.
    await hub.bus.publish("family_changed", {"collection": "reminders"})
